[mlir][tosa] Fix check for isolated regions in `tosa.cond_if`

lhutton1 opened the pull request:
This commit fixes a check in the validation pass which intended to validate whether a `tosa.cond_if` operation was conformant to the specification. The specification requires all values used in the then/else regions are explicitly declared within the regions. This change checks that these regions are 'isolated from above', to ensure this requirement is true.

lhutton1 committed:
[mlir][tosa] Fix check for isolated regions in `tosa.cond_if`

This commit fixes a check in the validation pass which intended
to validate whether a `tosa.cond_if` operation was conformant to
the specification. The specification requires all values used in
the then/else regions are explicitly declared within the regions.
This change checks that these regions are 'isolated from above',
to ensure this requirement is true.

Change-Id: I1b6eac1ed571e6b1eda4a58f0677c80e22977e58

FranklandJack commented on mlir/test/Dialect/Tosa/error_if_check.mlir:
```diff
@@ -227,15 +227,41 @@ func.func @test_error_i32_unsigned_output(%arg0: tensor<1xi8>) -> tensor<1xi32>
+
+// -----
+
+// COM: Check isolated cond_if's are valid
```
Does this need to use the `COM` directive? Unless I'm mistaken there aren't any strings in this comment that would match another FileCheck directive.

FranklandJack commented on mlir/lib/Dialect/Tosa/Transforms/TosaValidation.cpp:
```diff
@@ -1229,18 +1252,15 @@ bool checkErrorIfCondIf(Operation *op) {
-  bool isThenGraphNullaryRegion = isNullaryRegion(thenGraph);
-  bool isElseGraphNullaryRegion = isNullaryRegion(elseGraph);
-  bool isInputListEmpty = ifOp.getInputList().size() == 0;
+  bool isThenGraphIsolatedRegion = isIsolatedRegion(thenGraph);
```
Whilst we are here, could we make these bools `const`?

FranklandJack commented on mlir/test/Dialect/Tosa/error_if_check.mlir:
```diff
@@ -227,15 +227,41 @@ func.func @test_error_i32_unsigned_output(%arg0: tensor<1xi8>) -> tensor<1xi32>
-func.func @test_cond_if_simplified_form(%arg0: tensor<f32>, %arg1: tensor<f32>, %arg2: tensor<i1>) -> tensor<f32> {
-  // expected-error@+1 {{'tosa.cond_if' op the current simplified form is not strictly conformant to the spec, please use the generic format}}
+
+func.func @test_cond_if_not_isolated_from_above(%arg0: tensor<f32>, %arg1: tensor<f32>, %arg2: tensor<i1>) -> tensor<f32> {
```
If I understand correctly, this test (and the one below) have values in both the then and else blocks that are not isolated from above? Should we add a second test and check each block seperately?

FranklandJack commented on mlir/lib/Dialect/Tosa/Transforms/TosaValidation.cpp:
```diff
@@ -1229,18 +1252,15 @@ bool checkErrorIfCondIf(Operation *op) {
     op->emitOpError()
-        << "the current simplified form is not strictly conformant to the "
-           "spec, please use the generic format\n";
+        << "is not conformant to the TOSA specification. It requires the "
```
Weak preference for testing each condition seperately so we can add more info to the error message e.g. `requires the "then" region be isolated from above`.

FranklandJack reviewed (changes requested):
LGTM % some rather pedantic comments 😁

FranklandJack commented on mlir/lib/Dialect/Tosa/Transforms/TosaValidation.cpp:
```diff
@@ -1193,32 +1193,55 @@ bool checkErrorIfPad(Operation *op) {
+  //
+  // Generic:
+  // %0 = "tosa.cond_if"(%arg0, %arg1, %arg2) ({
+  //   ^bb0(%arg3, %arg4):
```
Both blocks here use the label `bb0`, is that expected? This is also the case for the lit tests below.

udaya-ranga commented on mlir/test/Dialect/Tosa/error_if_check.mlir:
```diff
@@ -227,15 +227,41 @@ func.func @test_error_i32_unsigned_output(%arg0: tensor<1xi8>) -> tensor<1xi32>
+// COM: Check isolated cond_if's are valid
+func.func @test_cond_if_isolated_from_above(%arg0: tensor<f32>, %arg1: tensor<f32>, %arg2: tensor<i1>) -> tensor<f32> {
+  %0 = "tosa.cond_if"(%arg2, %arg0, %arg1) ({
+    ^bb0(%arg3: tensor<f32>, %arg4: tensor<f32>):
```
What is relation of `arg3` to `(%arg2, %arg0, %arg1)` ?

lhutton1 commented on mlir/lib/Dialect/Tosa/Transforms/TosaValidation.cpp:
```diff
@@ -1193,32 +1193,55 @@ bool checkErrorIfPad(Operation *op) {
+static bool isOpIsolatedFromAbove(Operation *op, Region *region) {
+  return llvm::all_of(op->getOperands(), [&](auto operand) {
+    Region *operandRegion = operand.getParentRegion();
+    return region->isAncestor(operandRegion);
```
> Does "ancestor" relationship always equate to "IsolatedFromAbove" ?

Yes as far as I'm aware. Infact, this is the same style of check used in the `IsolatedFromAbove` trait verifier: https://github.com/llvm/llvm-project/blob/89c5c3ba38492de8163c344d076733b32d522727/mlir/lib/IR/Operation.cpp#L1369

lhutton1 committed:
Address review comments

- check then/else region isolation independently
- renamed isOpIsolatedFromAbove to isOpIsolatedWithinRegion
- remove "COM" prefix in lit test

Change-Id: I5988ba2e75c9aa81c57321628d49d35634847f03

lhutton1 reviewed (commented):
Thanks for the reviews @FranklandJack, @udaya-ranga!

udaya-ranga reviewed (commented):
LGTM

FranklandJack commented:
LGTM!
